File: meme_inator_back/apps/profiles/application/usecases/get_my_profile_usecase.py
from typing import List, Optional
from uuid import UUID
from apps.profiles.domain.entities.profile_entity import ProfileEntity
from apps.profiles.domain.entities.profile_light_entity import ProfileLightEntity
from apps.profiles.domain.usecases.iget_my_profile_usecase import IGetMyProfileUsecase
from apps.profiles.domain.irepositories.i_profile_repository import IProfileRepository
from core.results import Result

class GetMyProfileUsecase(IGetMyProfileUsecase):
    def __init__(self, profile_repo: IProfileRepository):
        self.profile_repo = profile_repo

    def execute(
            self, 
            user_id: UUID, 
            viewer_user_id: Optional[UUID] = None,
            fields: Optional[List[str]] = None
        ) -> Result[ProfileEntity|ProfileLightEntity]:
        # Delegate directly to repository – it already returns a Result
        return self.profile_repo.get_my_profile(user_id, viewer_user_id, fields)
    

File: meme_inator_front/lib/features/profiles/domain/usecases/get_my_profile_usecase.dart
import 'package:meme_inator_front/core/results.dart';
import 'package:meme_inator_front/features/profiles/domain/entities/profile_entity.dart';
import 'package:meme_inator_front/features/profiles/domain/interfaces/iprofile_repository.dart';

/// Simple use case to fetch the current user's own profile
class GetMyProfileUsecase {
  final IProfileRepository repository;

  GetMyProfileUsecase({required this.repository});

  /// Execute the use case
  Future<Result<ProfileEntity>> execute() async {
    try {
      return await repository.getMyProfile();
    } catch (e) {
      return Error.fromException(
        e as Exception,
        message: 'Failed to get my profile: ${e.toString()}',
      );
    }
  }
}